FTPFolderUploader/main.py

```diff
@@ -0,0 +1,293 @@
+                                        walkdir)
+                if not is_success:
+                    break
+
+            else:
+                print('.')
+
+            last_files_list = latest_files_list[:]  # copy the list to hold
+            time.sleep(SLEEP_SECONDS)
+
+        except(KeyboardInterrupt):
+            print()
+            print('Exiting.')
+            break
+
+
+def upload_all(server,
+               username,
+               password,
+               local_dir,
+               files_to_update=None,
+               walkdir=False
+               ):
+    print('\nUploading files in (%s)' % (os.path.abspath(local_dir)))
+    continue_on = False
+    server_connect_ok = False
+
+    local_dir = os.path.abspath(local_dir)
+
+    if files_to_update:
+        local_files = files_to_update
+    else:
+        local_files = _get_local_files(local_dir, walkdir)
+
+    if local_files:
+        try:
+            connection = paramiko.Transport(server,22)
+            # todo Authenication method ?
+            connection.connect(username=username, password=password)
+            sftp = paramiko.SFTPClient.from_transport(connection)
+
+            print('Logged in successfully (%s) as (%s)' % (server, username))
+
+            server_connect_ok = True
+
+        except paramiko.ssh_exception.SSHException as e:
+            print(e)
+        except paramiko.ssh_exception.NoValidConnectionsError as e:
+            print('SSH transport is not ready. %s' % (str(e.args)))
+
+        if server_connect_ok:
+
+            sftp.chdir(REMOTE_INCOMING_DIR)
+            for file_info in local_files:
+
+                source = file_info['path']
+                path, filename = os.path.split(source)
+                remote_path = source.replace(local_dir, '')
+                remote_path = remote_path.replace('\\', '/') # Convert to unix style and strip
+                relative_remote_path = remote_path.strip("/")
+
+                if os.path.exists(path):
+                    try:
+
+                        # todo compare mtimes with server
+                        mkdir_p(sftp, remote_path.replace(filename, ''))
+                        response = sftp.put(source, relative_remote_path, callback=None, confirm=True)
+                        # todo response contains atime and mtime. Store this ?
+                        print("Uploaded/checked %s" % relative_remote_path)
+                        continue_on = True
+                    except Exception as e:
+                        print('ERROR!')
+                        print(str(e.args))
+                        print()
+                else:
+                    print("WARNING -- File no longer exists, (%s)!" % source)
+
+            connection.close()
+            print('Closing Connection')
+    else:
+        print('ERROR -- No files found in (%s) walkdir=%s' % (local_dir, walkdir))
+
+    return continue_on
+
+
+def mkdir_p(sftp, remote_directory):
```

Comment:
I totally forgot that I had some Python SFTP code as part of the Pipeline project... some of it is a bit Pipeline specific, but others like the mkdir_p stuff was modelled on the standard library local makedir function, and is heavily unit tested there, so could be a good drop in replacement:

https://github.com/aodn/python-aodncore/blob/master/aodncore/pipeline/storage.py#L355